src/Agent.Listener/SelfUpdater.cs

```diff
@@ -141,95 +141,150 @@ private async Task DownloadLatestAgent(CancellationToken token)
+                        agentSuffix++;
                     }
                 }
 
-                if (archiveFile.EndsWith(".zip", StringComparison.OrdinalIgnoreCase))
-                {
-                    ZipFile.ExtractToDirectory(archiveFile, latestAgentDirectory);
-                }
-                else if (archiveFile.EndsWith(".tar.gz", StringComparison.OrdinalIgnoreCase))
+                Trace.Info($"Attempt {attempt}: save latest agent into {archiveFile}.");
+
+                bool gettingPackageSucceeded = false;
+
+                try
                 {
-                    string tar = WhichUtil.Which("tar", trace: Trace);
-                    if (string.IsNullOrEmpty(tar))
-                    {
-                        throw new NotSupportedException($"tar -xzf");
-                    }
+                    bool downloadSucceeded = false;
 
-                    // tar -xzf
-                    using (var processInvoker = HostContext.CreateService<IProcessInvoker>())
+                    using (var httpClient = new HttpClient(HostContext.CreateHttpClientHandler()))
                     {
-                        processInvoker.OutputDataReceived += new EventHandler<ProcessDataReceivedEventArgs>((sender, args) =>
+                        // Set a 15-minute timeout per agent package download attempt
+                        httpClient.Timeout = TimeSpan.FromMinutes(15);
+
+                        try
                         {
-                            if (!string.IsNullOrEmpty(args.Data))
+                            //open zip stream in async mode
+                            using (FileStream fs = new FileStream(archiveFile, FileMode.Create, FileAccess.Write, FileShare.None, bufferSize: 4096, useAsync: true))
                             {
-                                Trace.Info(args.Data);
+                                using (Stream result = await httpClient.GetStreamAsync(_targetPackage.DownloadUrl))
+                                {
+                                    //81920 is the default used by System.IO.Stream.CopyTo and is under the large object heap threshold (85k). 
+                                    await result.CopyToAsync(fs, 81920, token);
+                                    await fs.FlushAsync(token);
+                                }
                             }
-                        });
 
-                        processInvoker.ErrorDataReceived += new EventHandler<ProcessDataReceivedEventArgs>((sender, args) =>
+                            downloadSucceeded = true;
+                        }
+                        catch (Exception ex)
+                        {
+                            Trace.Warning($"Download: failed to get package '{archiveFile}' from CDN. Exception {ex}");
+                        }
+                    }
+
+                    if (downloadSucceeded)
```

Comment:
should the unzip be outside of the retry look? i think we might only need retry around download right now.